### PR TITLE
[IoT Hub] Support for 2020-03-01 API and Network Isolation commands

### PR DESCRIPTION
**Pull request by c-ryan-k**
**Description**  

- Support for new IoT Hub 2020-03-01 API
- SDK update for azure-mgmt-iot-hub to 0.12.0
- `az routing-endpoint-create` auth-type support
- File Upload auth-type support

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
IoT Hub

File: src/azure-cli/azure/cli/command_modules/iot/_help.py [hunk 1]
```diff
@@ -968,3 +971,70 @@
 type: command
 short-summary: Update metadata for an IoT Central application.
 """
+
+helps['iot hub private-endpoint-connection'] = """
+type: group
+short-summary: Manage private endpoint connections to an IoT Hub.
+"""
+
+helps['iot hub private-endpoint-connection approve'] = """
```
Comment (MyronFanQiu) [inline src/azure-cli/azure/cli/command_modules/iot/_help.py:980]: Please implement private endpoint connection and private link resource command under network. You can simply register the type, api-version and add corresponding tests. 
https://github.com/Azure/azure-cli/blob/fe18bc1c9832da8f0d1dc9082d14bbcd800926e5/src/azure-cli/azure/cli/command_modules/network/private_link_resource_and_endpoint_connections/custom.py#L15-L22

Comment (MyronFanQiu) [reply src/azure-cli/azure/cli/command_modules/iot/_help.py:980]: our official doc is a little bit out-of-date. We will change it later. If you have any question, please reach out to me.

File: src/azure-cli/azure/cli/command_modules/iot/custom.py [hunk 2]
```diff
@@ -398,6 +400,11 @@ def iot_hub_create(cmd, client, hub_name, resource_group_name, location=None,
         raise CLIError('Please mention storage container name.')
     if fileupload_storage_container_name and not fileupload_storage_connectionstring:
         raise CLIError('Please mention storage connection string.')
+    identityBased_fileUpload = fileupload_storage_authentication_type and fileupload_storage_authentication_type.lower() == AuthenticationType.IdentityBased.value
```
Comment (MyronFanQiu) [inline src/azure-cli/azure/cli/command_modules/iot/custom.py:403]: please use pythonic style. `identity_based_file_upload`

File: src/azure-cli/azure/cli/command_modules/iot/custom.py [hunk 3]
```diff
@@ -499,6 +510,15 @@ def update_iot_hub_custom(instance,
     if fileupload_notification_ttl is not None:
         ttl = timedelta(hours=fileupload_notification_ttl)
         instance.properties.messaging_endpoints['fileNotifications'].ttl_as_iso8601 = ttl
+
+    identityBased_fileUpload = fileupload_storage_authentication_type and fileupload_storage_authentication_type.lower() == AuthenticationType.IdentityBased.value
+    if identityBased_fileUpload:
+        instance.properties.storage_endpoints['$default'].authentication_type = AuthenticationType.IdentityBased
+        instance.properties.storage_endpoints['$default'].container_uri = fileupload_storage_container_uri
+    else:
```
Comment (MyronFanQiu) [inline src/azure-cli/azure/cli/command_modules/iot/custom.py:518]: We should check whether the fileupload_storage_authentication_type is None or not. User just doesn't want to change anything. According to current logic, the container_uri would be set to None instead.

File: src/azure-cli/azure/cli/command_modules/iot/custom.py [hunk 4]
```diff
@@ -418,7 +425,9 @@ def iot_hub_create(cmd, client, hub_name, resource_group_name, location=None,
     storage_endpoint_dic['$default'] = StorageEndpointProperties(
         sas_ttl_as_iso8601=timedelta(hours=fileupload_sas_ttl),
         connection_string=fileupload_storage_connectionstring if fileupload_storage_connectionstring else '',
-        container_name=fileupload_storage_container_name if fileupload_storage_container_name else '')
+        container_name=fileupload_storage_container_name if fileupload_storage_container_name else '',
```
Comment (MyronFanQiu) [inline src/azure-cli/azure/cli/command_modules/iot/custom.py:428]: should it be `fileupload_storage_container_name if fileupload_storage_container_name else None`? What's the difference between `None` and `""`?

Comment (c-ryan-k) [reply src/azure-cli/azure/cli/command_modules/iot/custom.py:428]: To clarify, would you rather them all be `None`, or just `fileupload_storage_container_name` ?
Was just trying to change as little as possible in existing code when adding a new parameter, it had existed as `''` before

Comment (c-ryan-k) [reply src/azure-cli/azure/cli/command_modules/iot/custom.py:428]: In reality I don't believe there's a difference in result on the service end, but I could be wrong.

File: src/azure-cli/HISTORY.rst [hunk 5]
```diff
@@ -75,6 +75,14 @@ Release History
 * `az vm availability-set list`: Support showing VM list (#13090)
 * `az vm list-skus`: Fix display problem of table format (#13184)
 
+**IoT**
```
Comment (MyronFanQiu) [inline src/azure-cli/HISTORY.rst:78]: We don't directly change this file anymore. Please modify your title and first comment based on the instruction to add history.

File: src/azure-cli/azure/cli/command_modules/iot/_validators.py [hunk 6]
```diff
@@ -75,3 +76,20 @@ def validate_c2d_ttl(ns):
             ns.c2d_ttl not in range(1, 49, 1)):
         raise ArgumentError(None, 'Please specify the default time from 1 to 48 hours to live for cloud-to-device'
                                   ' messages in the device queue.')
+
+
+def validate_private_endpoint_connection_id(namespace):
```
Comment (MyronFanQiu) [inline src/azure-cli/azure/cli/command_modules/iot/_validators.py:81]: do we still need this function?

File: src/azure-cli/azure/cli/command_modules/iot/custom.py [hunk 7]
```diff
@@ -672,11 +692,26 @@ def iot_hub_get_stats(client, hub_name, resource_group_name=None):
...
+    if EndpointType.AzureStorageContainer.value == endpoint_type.lower() and not has_endpoint_uri:
+        raise CLIError("Please provide an endpoint uri '--endpoint-uri'")
+    if not has_endpoint_uri_and_path:
+        raise CLIError("Please provide an endpoint uri '--endpoint-uri' and entity path '--entity-path'")
+
+
 def iot_hub_routing_endpoint_create(cmd, client, hub_name, endpoint_name, endpoint_type,
                                     endpoint_resource_group, endpoint_subscription_id,
-                                    connection_string, container_name=None, encoding=None,
+                                    connection_string=None, container_name=None, encoding=None,
```
Comment (MyronFanQiu) [inline src/azure-cli/azure/cli/command_modules/iot/custom.py:711]: connection_string is not required anymore?

File: src/azure-cli/azure/cli/command_modules/iot/tests/latest/test_iot_commands.py [hunk 8]
```diff
@@ -395,6 +395,120 @@ def test_iot_hub(self, resource_group, resource_group_location, storage_account)
...
+                 checks=[self.check('properties.privateLinkServiceConnectionState.status', 'Approved')])
+
+        # endpoint connection reject by name
+        reject_desc = 'Rejecting endpoint connection'
+        self.cmd('network private-endpoint-connection reject --type {0} -n {1} --resource-name {2} -g {3} --description "{4}"'
+                 .format(private_endpoint_type, private_endpoint_name, identity_hub, rg, reject_desc),
+                 checks=[self.check('properties.privateLinkServiceConnectionState.status', 'Rejected')])
+
+        # endpoint show
+        self.cmd('network private-endpoint-connection show --type {0} -n {1} --resource-name {2} -g {3}'
```
Comment (MyronFanQiu) [inline src/azure-cli/azure/cli/command_modules/iot/tests/latest/test_iot_commands.py:504]: could we test `az network private-endpoint-connection list` as well?

File: src/azure-cli/azure/cli/command_modules/iot/_validators.py (hunk 6 above)
Comment (c-ryan-k) [reply src/azure-cli/azure/cli/command_modules/iot/_validators.py:81]: Nope, must have missed that in the network code refactor. Will remove.

File: src/azure-cli/azure/cli/command_modules/iot/custom.py (hunk 7 above)
Comment (c-ryan-k) [reply src/azure-cli/azure/cli/command_modules/iot/custom.py:711]: Correct - if the routing-endpoint uses `identityBased` authentication, it does not use a connection string.

**Comment by MyronFanQiu** [comment]
@fengzhou-msft @jiasli Could you review this PR since you're code owner too.

**Pull request merged**